### PR TITLE
Add concurrent iterator functions

### PR DESCRIPTION
**Pull request by kelunik**
This allows simplifying https://github.com/Insolita/php-async-benchmarks/pull/2.

```php
private function processRequests(Iterator $urls): Promise
{
    return call(function () use ($urls) {
        $this->goodFile = yield File\open($this->tempDir . '/ok.txt', 'a');
        $this->badFile = yield File\open($this->tempDir . '/bad.txt', 'a');

        try {
            yield Queue::fromIterator($urls)->process(new LocalSemaphore($this->concurrency),
                function (string $url) {
                    /** @var Response $response */
                    $response = yield $this->client->request(new Request($url));
                    yield $this->processHtml(yield $response->getBody()->buffer(), $url);
                }, function (\Throwable $error, string $url) {
                    yield $this->badFile->write($url . \PHP_EOL);
                }
            );
        } finally {
            yield $this->goodFile->close();
            yield $this->badFile->close();
        }
    });
}

### CODE REVIEW

File: src/Queue.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+                $this->pending[$promiseId] = $promise;
+                $promise->onResolve(function () use ($promiseId) {
+                    unset($this->pending[$promiseId]);
+                });
+            }
+
+            yield Promise\any($this->pending);
+        });
+    }
+}
```
Comment (enumag) [inline src/Queue.php:67]:
```suggestion
}
```

**Comment by Insolita** [comment]
@kelunik this is good for my example task when we have a predefined set of urls, but what about a crawling task, when we will want add to queue additional urls from a response handler. Is it will be not overcomplicated if we will wrap links scraped from page to new producer and call processRequests() recursively?

File: src/Queue.php (hunk 1 above)
Comment (enumag) [inline src/Queue.php:67]: add newline?

**Comment by kelunik** [comment]
@Insolita It shouldn't use recursion, no. You can use `Amp\Emitter` instead, the only thing you need to know is when to call `Emitter::complete()`. I've added an example, do you think that's fine?

File: src/Queue.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    /** @var Iterator */
+    private $iterator;
+
+    /** @var bool */
+    private $processing = false;
+
+    /** @var Promise[] */
+    private $pending = [];
```
Comment (enumag) [inline src/Queue.php:24]: Perhaps use SplObjectStorage instead?

**Comment by trowski** [comment]
@kelunik How about rolling this into a function returning an `Iterator`. Since an iterator can only be transversed once, holding it in an object doesn't really make sense IMO.

```php
function concurrentMap(Iterator $iterator, Semaphore $semaphore, callable $processor): Iterator
{
    return new Producer(function (callable $emit) use ($iterator, $semaphore, $processor): \Generator {
        $pending = [];

        while (yield $iterator->advance()) {
            /** @var Lock $lock */
            $lock = yield $semaphore->acquire();

            $job = $iterator->getCurrent();

            $promise = call(function () use ($lock, $job, $processor, $emit): \Generator {
                try {
                    $value = yield call($processor, $job);
                } finally {
                    $lock->release();
                }

                yield $emit($value);
            });

            $promiseId = \spl_object_id($promise);

            $pending[$promiseId] = $promise;
            $promise->onResolve(function () use (&$pending, $promiseId): void {
                unset($pending[$promiseId]);
            });
        }

        yield Promise\any($pending);
    });
}
```

**Comment by kelunik** [comment]
@trowski Seems fine, too, but we still need the error handler, otherwise you can only react to the first failure.

File: src/Queue.php (hunk 2 above)
Comment (kelunik) [reply src/Queue.php:24]: That'd create a circular reference. :-(

**Comment by trowski** [comment]
> @trowski Seems fine, too, but we still need the error handler, otherwise you can only react to the first failure.

I guess that depends. `Amp\Iterator\map()` fails the returned `Iterator` the first time the callback throws. So I'd say the same should happen here.

File: src/functions.php [hunk 3]
```diff
@@ -29,3 +32,68 @@ function synchronized(Mutex $mutex, callable $callback, ...$args): Promise
...
+                $processor,
+                $emit,
+                &$error
+            ) {
+                try {
+                    yield $emit(yield $processor($currentElement));
+                } catch (\Throwable $e) {
+                    if ($error === null) {
+                        $error = $e;
+                    }
```
Comment (enumag) [inline src/functions.php:79]: Is this necessary? I'm not sure but I think the effect might be the same if you don't catch at all (just keep the finally)

Comment (kelunik) [reply src/functions.php:79]: In that case `$error` is never set and it never fails?

Comment (enumag) [reply src/functions.php:79]: I meant remove the $error variable completely too... just try if the tests pass like that, if not then I missed something

Comment (kelunik) [reply src/functions.php:79]: No, as `$promise` is never yielded or errors otherwise handled.

**Comment by kelunik** [comment]
Got it working with correct GC behavior \o/

File: src/functions.php [hunk 4]
```diff
@@ -29,3 +33,158 @@ function synchronized(Mutex $mutex, callable $callback, ...$args): Promise
...
+ *     otherwise the entire operation is aborted.
+ *
+ * @return Promise
+ */
+function concurrentForeach(Iterator $iterator, Semaphore $semaphore, callable $processor): Promise
+{
+    $processor = coroutine($processor);
+
+    return call(static function () use ($iterator, $semaphore, $processor) {
+        yield Iterator\toArray(concurrent($iterator, $semaphore, coroutine(static function ($value) use ($processor) {
```
Comment (trowski) [inline src/functions.php:186]: Consider adding `return` here. Couldn't hurt, right?

Comment (kelunik) [reply src/functions.php:186]: I need it, because the return value shouldn't always be an empty array, but rather `Promise<void>`

**Pull request merged**